### DISCUSSION TITLE
php sessions in redis

### DISCUSSION THREAD

**Discussion by brendankearney** [ampache/ampache#3358] [General]
i recently stood up a Redis footprint with Sentinel for HA/Replication for other PHP apps i am running to store session data in Redis so that i can load balance across multiple HTTPd instances, while maintaining the logged in session state.  this has greatly improved performance in the other apps, as i can truly load balance across the web servers, instead of persisting a session to a single box.  in HAProxy i can use "leastconn" load balancing instead of "source-ip", spreading the requests across both of the web servers.

i see that the local_web_path variable can be set to the proper URL when behind a reverse proxy or load balancer like HAProxy.  being that i am using HAProxy, but have my ampache instance running on my nas, could i run disparate ampache instances on my web servers, leveraging HAProxy for load balancing and Redis for session sharing without breaking the catalogs?

the php and php-fpm configs would be pointed to redis using the php-pecl-redis5 package (on fedora), and the session data would written to / read from the active redis server.  the web servers would have an automount destination set for the NFS share on the nas, and would have the same mountpoint and access to the music library, so there would not be conflict or contention in accessing the data. i am just unsure if there is any other "gotcha" or pitfall in setting ampache up like this.

**Comment by Andrew-McGee** [comment]
What about the database? How would you keep it synced across servers? Or you have one db that both ampache instances connect to?

**Comment by brendankearney** [comment]
i have a 3 node, multi-primary MariaDB footprint setup behind HAProxy, using Galera replication.  all MariaDB instances are read/write and replicate all actions to the other cluster members, including the MyISAM tables, via the experimemtal "wsrep_replicate_myisam" directive.  even now, i point my ampache instance at the HAProxy VIP for database access, so this would be more of the same, except it would be the web servers hosting the ampache app and pointing at the same HAProxy VIP for database access.

**Reply by Andrew-McGee** [reply]
Nice set up. I have no advice for you as I have not worked with HA clusters for many years so my knowledge is not up to date. But if you get something working I would like to hear how it works - mostly for scaling/load balancing. Also if you notice any performance hit from the NFS mounted library or the replicated db. Assume this is not a stretch cluster (it's all local nodes and not geo dispersed).

**Comment by brendankearney** [comment]
i'll keep this discussion updated with progress.  the hosts involved are server1, server2 and nas... very pragmatic naming scheme... they are all on the same subnet and are right next to each other.  i am not expecting a significant hit in performance, as i have done my homework on performance tuning, or at least tried to. :D

server1 and server2 are going to be the hosting web servers, instead of nas currently.
server1 server2 and nas are all the participating members in the mariadb cluster.
server1 and server2 will have a NFS automount for the music files in the catalog, and the mountpoint will be identical for ease of administration and the like.
the php session info will be stored in the redis KVP database that server1, server2 and nas participate in as well.

**Comment by brendankearney** [comment]
i had to export the music shares specifically for server1 and server2, since the only other export of the NFS share was for a user that had kerberized access required (krb5p).  after exporting the NFS share for the servers without an authentication requirement, i could add the catalog with the automount path set.  so far things seem to work, though a couple odd behaviors have been encountered.

upon login, the page is white and a refresh (press F5), and resend the POSTed data clears the "event".  oddly, i also do not see any session information being stored in Redis, though php and php-fpm are configured to use Redis with "session.save_handler = redis" and "session.save_path = "tcp://redis.bpk2.com:6379?auth=blahblahblah"

**Comment by brendankearney** [comment]
i found that when i was using both server1 and server2 as the web servers, not nas, when i added songs to the html5 web player playlist, the file would not play.  it would attempt to play the file, but its like the file was 0.00 seconds long and the blue bar instantly went all the way across the track duration bar.  its seems that something with the track selection and track playback is not in sync?
